### PR TITLE
Added support for initializing Collection using a constructor

### PR DESCRIPTION
**Pull request by mlewand**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Added support for initializing `Collection` items using a constructor. Closes ckeditor/ckeditor5#6319.

---

### Additional information

This PR was mostly based on #309 with some notable changes:

*   Reverted `collection-add-invalid-id` and `collection-add-item-already-exists` error interfaces back to what we have on the `master` branch (see [current](6f51d49026045ef9919cccab68eeeee736c4c7ac) vs [originally proposed](https://github.com/ckeditor/ckeditor5-utils/blob/1aac49db1ca9ea914f9fc44f00e295474faf3ed2/src/collection.js#L669-L693)). I don't think there's a strong reason to change this API.
*   `add()` remains unchanged with exception of reusing `_getItemIdBeforeAdding()`.
*   Added API docs.
*   Added couple more test cases: `generates id for items that doesn\'t have it`, `throws an error when invalid item key is provided`, `throws an error when items have a duplicated key`.

I had some doubts:

*   I wonder whether it would be better to do the item initialization as `options.items` instead of [adding `initialItemsOrOptions` next to `options`](https://github.com/ckeditor/ckeditor5-utils/blob/58bd45bcee5384f0e217f66063a1d5bc58eb5ce2/src/collection.js#L64). I believe any `xOrY` kind of variables are confusing for the API consumer.

Once this PR is merged, the `master` branch should get merged to the [#309](https://github.com/ckeditor/ckeditor5-utils/pull/309) PR so that the diff will get smaller and easier to review.

There's a subpr: ckeditor/ckeditor5-ui#545

### CODE REVIEW

File: src/collection.js [hunk 1]
```diff
@@ -28,10 +29,45 @@ export default class Collection {
...
+	 * Whatever option you choose, you can always pass a configuration object as the last argument
+	 * of the constructor:
+	 *
+	 *		const emptyCollection = new Collection( { idProperty: 'name' } );
+	 *		emptyCollection.add( { name: 'John' } );
+	 *		console.log( collection.get( 'John' ) ); // -> { name: 'John' }
+	 *
+	 *		const nonEmptyCollection = new Collection( [ { name: 'John' } ], { idProperty: 'name' } );
+	 *		nonEmptyCollection.add( { name: 'George' } );
+	 *		console.log( collection.get( 'George' ) ); // -> { name: 'George' }
```
Comment (Reinmar) [inline src/collection.js]:
```suggestion
	 *		console.log( collection.get( 'George' ) ); // -> { name: 'George' }
	 *		console.log( collection.get( 'John' ) ); // -> { name: 'John' }
```

File: src/collection.js [hunk 2]
```diff
@@ -28,10 +29,45 @@ export default class Collection {
 	/**
 	 * Creates a new Collection instance.
 	 *
-	 * @param {Object} [options={}] The options object.
-	 * @param {String} [options.idProperty='id'] The name of the property which is considered to identify an item.
+	 * You can provide an array of initial items the collection will be created with:
```
Comment (oleq) [inline src/collection.js]: It's no longer an array. It's iterable.

File: tests/collection.js [hunk 3]
```diff
@@ -25,18 +25,83 @@ describe( 'Collection', () => {
 	} );
 
 	describe( 'constructor()', () => {
-		it( 'allows to change the id property used by the collection', () => {
-			const item1 = { id: 'foo', name: 'xx' };
-			const item2 = { id: 'foo', name: 'yy' };
-			const collection = new Collection( { idProperty: 'name' } );
+		describe( 'setting initial collection items', () => {
+			it( 'works using an array', () => {
```
Comment (oleq) [inline tests/collection.js]:
```suggestion
			it( 'should work using an array', () => {
```

File: tests/collection.js [hunk 4]
```diff
@@ -25,18 +25,83 @@ describe( 'Collection', () => {
...
+				expect( collection ).to.have.length( 2 );
 
-			expect( collection ).to.have.length( 2 );
+				expect( collection.get( 0 ) ).to.equal( item1 );
+				expect( collection.get( 1 ) ).to.equal( item2 );
+				expect( collection.get( 'foo' ) ).to.equal( item1 );
+				expect( collection.get( 'bar' ) ).to.equal( item2 );
+			} );
+
+			it( 'works using an iterable', () => {
```
Comment (oleq) [inline tests/collection.js]:
```suggestion
			it( 'should work using an iterable', () => {
```

File: tests/collection.js [hunk 5]
```diff
@@ -25,18 +25,83 @@ describe( 'Collection', () => {
...
+				const itemsSet = new Set( [ item1, item2 ] );
+				const collection = new Collection( itemsSet );
+
+				expect( collection ).to.have.length( 2 );
+
+				expect( collection.get( 0 ) ).to.equal( item1 );
+				expect( collection.get( 1 ) ).to.equal( item2 );
+			} );
+
+			it( 'generates id for items that doesn\'t have it', () => {
```
Comment (oleq) [inline tests/collection.js]:
```suggestion
			it( 'should generate ids for items that doesn\'t have it', () => {
```

File: tests/collection.js [hunk 6]
```diff
@@ -25,18 +25,83 @@ describe( 'Collection', () => {
...
+				const item = {};
+				const collection = new Collection( [ item ] );
+
+				expect( collection.get( 0 ).id ).to.be.a( 'string' );
+				expect( collection.get( 0 ).id ).not.to.be.empty;
+			} );
+
+			it( 'throws an error when invalid item key is provided', () => {
```
Comment (oleq) [inline tests/collection.js]:
```suggestion
			it( 'should throw an error when an invalid item key is provided', () => {
```

File: tests/collection.js [hunk 7]
```diff
@@ -25,18 +25,83 @@ describe( 'Collection', () => {
...
+				const badIdItem = getItem( 1 ); // Number id is not supported.
+
+				expectToThrowCKEditorError( () => {
+					return new Collection( [ badIdItem ] );
+				}, /^collection-add-invalid-id/ );
+			} );
+
+			it( 'throws an error when items have a duplicated key', () => {
```
Comment (oleq) [inline tests/collection.js]:
```suggestion
			it( 'should throw an error when two items have the same key', () => {
```

File: tests/collection.js [hunk 8]
```diff
@@ -25,18 +25,83 @@ describe( 'Collection', () => {
...
+				const item2 = getItem( 'foo' );
+
+				expectToThrowCKEditorError( () => {
+					return new Collection( [ item1, item2 ] );
+				}, /^collection-add-item-already-exists/ );
+			} );
+		} );
+
+		describe( 'options', () => {
+			it( 'allow to change the id property used by the collection', () => {
```
Comment (oleq) [inline tests/collection.js]:
```suggestion
			it( 'should allow to change the id property used by the collection', () => {
```

File: tests/collection.js [hunk 9]
```diff
@@ -25,18 +25,83 @@ describe( 'Collection', () => {
...
 
-			expect( collection.get( 'xx' ) ).to.equal( item1 );
-			expect( collection.remove( 'yy' ) ).to.equal( item2 );
+				expect( collection ).to.have.length( 2 );
+
+				expect( collection.get( 'xx' ) ).to.equal( item1 );
+				expect( collection.remove( 'yy' ) ).to.equal( item2 );
+			} );
+
+			it( 'allow to change the id property used by the collection (initial items)', () => {
```
Comment (oleq) [inline tests/collection.js]:
```suggestion
			it( 'should allow to change the id property used by the collection (initial items passed to the constructor)', () => {
```

**Comment by oleq** [comment]
Let's merge it ASAP in the upcoming iteration.

**Pull request merged**